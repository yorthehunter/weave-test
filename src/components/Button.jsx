import React from 'react';
import { node, oneOf, bool } from 'prop-types';
import classNames from 'classnames';
// import 'weave-ui-button';
import '../weave/components/button/index.scss';

const componentPrefix = 'weave-button';

const Button = (props) => {
  const {
    type,
    size,
    responsive,
    disabled,
    fullWidth,
    children,
  } = props;

  const componentClasses = classNames([
    componentPrefix,
    size && `${componentPrefix}--${size}`,
    type && `${componentPrefix}--${type}`,
    responsive && `${componentPrefix}--responsive`,
    fullWidth && `${componentPrefix}--full-width`,
    disabled && 'is-disabled',
  ]);

  return (
    <button className={componentClasses}>
      {children}
    </button>
  );
};

Button.defaultProps = {
  type: undefined,
  size: undefined,
  disabled: false,
  responsive: false,
  fullWidth: false,
};

Button.propTypes = {
  type: oneOf([undefined, 'primary', 'secondary', 'cta', 'flat']),
  size: oneOf([undefined, 'xs', 'sm', 'md', 'lg']),
  disabled: bool,
  responsive: bool,
  fullWidth: bool,
  children: node.isRequired,
};

export default Button;
