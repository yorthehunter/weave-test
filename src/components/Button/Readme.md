Default Button
```js
<Button>Hello</Button>
```

Button styles
```js
<div>
  <Button type="primary">Hello</Button>
  <br />
  <Button type="secondary">Hello</Button>
  <br />
  <Button type="cta">Hello</Button>
  <br />
  <Button type="flat">Hello</Button>
</div>
```

Button sizes
```js
<div>
  <Button size="xs">Hello</Button>
  <br />
  <Button size="sm">Hello</Button>
  <br />
  <Button size="md">Hello</Button>
  <br />
  <Button size="lg">Hello</Button>
</div>
```

Submit button
```js
<Button submit>Hello</Button>
```
