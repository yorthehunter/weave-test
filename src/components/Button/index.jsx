import React from 'react';
import { oneOf, string, bool, func, node } from 'prop-types';
import { sizeModifier, buttonTypeModifier } from '../utilities/util';
import CustomTag from '../utilities/CustomTag';
import '../../weave/components/button/index.scss';

const componentPrefix = 'weave-button';

/**
 * Button renders a weave-button with size, type, and other optional modifiers.
 * Use buttons to cause actions, not to link between pages.
 */

const Button = (props) => {
  const {
    size,
    type,
    submit,
    tag,
    children,
    onClick,
    disabled,
    responsive,
    fullWidth,
  } = props;

  const componentClasses = [
    componentPrefix,
    size && `${componentPrefix}--${size}`,
    type && `${componentPrefix}--${type}`,
    responsive && `${componentPrefix}--responsive`,
    fullWidth && `${componentPrefix}--full-width`,
    disabled && 'is-disabled',
  ];

  return (
    <CustomTag
      {...props}
      tag={submit ? 'input' : tag}
      type={submit ? 'submit' : type}
      value={submit ? children : undefined}
      onClick={onClick}
      componentClasses={componentClasses}
    >
      { submit ? null : children }
    </CustomTag>
  );
};

const propTypes = {
  tag: string,
  target: string,
  type: oneOf([
    'primary',
    'secondary',
    'cta',
    'flat',
    null,
  ]),
  size: oneOf([
    'xs',
    'sm',
    'md',
    'lg',
    null,
  ]),
  /** If true, disables button */
  disabled: bool,
  /** If true, forces a width of 100% */
  fullWidth: bool,
  /** If true, creates a responsive button */
  responsive: bool,
  /** Generates a submit input element */
  submit: bool,
  /** What happens on click */
  onClick: func,
  children: node.isRequired,
};

const defaultProps = {
  tag: 'button',
  target: null,
  type: null,
  size: null,
  disabled: false,
  responsive: false,
  fullWidth: false,
  submit: false,
  onClick: undefined,
};

Button.propTypes = propTypes;
Button.defaultProps = defaultProps;
export default Button;
