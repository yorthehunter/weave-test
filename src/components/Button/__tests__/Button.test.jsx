import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Button from '../';

describe('Button component', () => {
  const buttonClass = '.weave-button';
  const wrapper = mount(<Button>Button</Button>);

  it('renders with default values', () => {
    const rendered = renderer.create(<Button>Button</Button>);
    expect(rendered.toJSON()).toMatchSnapshot();
  });

  it('handles disabled state', () => {
    wrapper.setProps({ isDisabled: true });
    expect(wrapper.find(buttonClass).hasClass('is-disabled')).toEqual(true);
  });

  it('renders as a submit button', () => {
    const rendered = renderer.create(<Button submit>Submit Button</Button>);
    expect(rendered.toJSON()).toMatchSnapshot();
  });

  it('correctly handles button type', () => {
    wrapper.setProps({ type: 'primary' });
    expect(wrapper.find(buttonClass).hasClass('weave-button--primary')).toEqual(true);
    wrapper.setProps({ type: 'secondary' });
    expect(wrapper.find(buttonClass).hasClass('weave-button--secondary')).toEqual(true);
    wrapper.setProps({ type: 'cta' });
    expect(wrapper.find(buttonClass).hasClass('weave-button--cta')).toEqual(true);
    wrapper.setProps({ type: 'flat' });
    expect(wrapper.find(buttonClass).hasClass('weave-button--flat')).toEqual(true);
    wrapper.setProps({ type: null });
    expect(wrapper.find(buttonClass).hasClass('weave-button--primary')).toEqual(false);
  });

  it('correctly handles button size', () => {
    wrapper.setProps({ size: 'sm' });
    expect(wrapper.find(buttonClass).hasClass('weave-button--sm')).toEqual(true);
    wrapper.setProps({ size: 'md' });
    expect(wrapper.find(buttonClass).hasClass('weave-button--md')).toEqual(true);
    wrapper.setProps({ size: 'lg' });
    expect(wrapper.find(buttonClass).hasClass('weave-button--lg')).toEqual(true);
    wrapper.setProps({ size: null });
    expect(wrapper.find(buttonClass).hasClass('weave-button--lg')).toEqual(false);
  });

  it('correctly renders as a link', () => {
    const props = {
      tag: 'a',
      href: 'testUrl',
      target: '_blank',
    };

    const rendered = renderer.create(<Button {...props}>Button</Button>);
    expect(rendered.toJSON()).toMatchSnapshot();
  });

  it('correctly renders when passed an onClick function', () => {
    const rendered = renderer.create(<Button onClick={jest.fn()}>Button</Button>);
    expect(rendered.toJSON()).toMatchSnapshot();
  });
});
