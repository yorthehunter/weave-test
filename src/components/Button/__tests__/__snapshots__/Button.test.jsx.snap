// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component correctly renders as a link 1`] = `
<a
  className="weave-button"
  href="testUrl"
  onClick={null}
  style={Object {}}
  type={null}
  value={null}
>
  Button
</a>
`;

exports[`Button component correctly renders when passed an onClick function 1`] = `
<button
  className="weave-button"
  href={null}
  onClick={[MockFunction]}
  style={Object {}}
  type={null}
  value={null}
>
  Button
</button>
`;

exports[`Button component renders as a submit button 1`] = `
<input
  className="weave-button"
  href={null}
  onClick={null}
  style={Object {}}
  type="submit"
  value="Submit Button"
/>
`;

exports[`Button component renders with default values 1`] = `
<button
  className="weave-button"
  href={null}
  onClick={null}
  style={Object {}}
  type={null}
  value={null}
>
  Button
</button>
`;
